jetstream/allocation.py

```diff
@@ -338,19 +354,58 @@ def fill_user_allocation_sources():
     return allocation_resources
 
 
-def fill_user_allocation_source_for(driver, user, force_update=True):
+def fill_user_allocation_source_for(driver, user):
+    from core.models import AtmosphereUser
+    assert isinstance(user, AtmosphereUser)
     allocation_list = find_user_allocation_source_for(driver, user)
     allocation_resources = []
+    user_allocation_sources = []
+    old_user_allocation_sources = list(UserAllocationSource.objects.filter(user=user).order_by(
+        'allocation_source__name').all())
+
     for api_allocation in allocation_list:
-        allocation_source, _ = get_or_create_allocation_source(
-            api_allocation, update_source=force_update)
-        resource, _ = UserAllocationSource.objects.get_or_create(
-            allocation_source=allocation_source,
-            user=user)
+        allocation_source = get_or_create_allocation_source(api_allocation)
         allocation_resources.append(allocation_source)
+        user_allocation_source = get_or_create_user_allocation_source(user, allocation_source)
+        user_allocation_sources.append(user_allocation_source)
+
+    canonical_source_names = [source.name for source in allocation_resources]
+    for user_allocation_source in old_user_allocation_sources:
+        if user_allocation_source.allocation_source.name not in canonical_source_names:
+            delete_user_allocation_source(user, user_allocation_source.allocation_source)
```

Comment:
This is beyond a "Tuesday" thing, but is it okay if a `user_allocation_source` within `old_user_allocation_sources` causes a failure that would _prevent_ other "user_allocation_source"s from being processed? 

If I have a list, like, say [asrc11, asrc12, asrc13, asrc14, asrc15, asrc16, asrc17] - and asrc12 caused a failure, then we don't know if asrc13-17 should be deleted? 

Maybe my question is gibberish.